### PR TITLE
Crud for agent roles

### PR DESCRIPTION
**Pull request by Raibo**
https://github.com/happy-travel/agent-app-project/issues/329

### CODE REVIEW

**Comment by Raibo** [comment]
I did not test it yet, because the table is empty, and I don't want to mess with Ids. I will test this after the data will be there.

File: Api/AdministratorServices/AgentRolesService.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+using System.Threading.Tasks;
+using FluentValidation;
+using HappyTravel.Edo.Api.AdministratorServices.Models;
+using HappyTravel.Edo.Api.Extensions;
+using HappyTravel.Edo.Api.Infrastructure;
+using HappyTravel.Edo.Data.Agents;
+
+namespace HappyTravel.Edo.Api.AdministratorServices
+{
+    public class AgentRolesService : IAgentRolesService
```
Comment (mingazhev) [inline Api/AdministratorServices/AgentRolesService.cs:16]: AgentRolesManagementService will be more clear, I think

File: Api/AdministratorServices/AgentRolesService.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                _context.Update(role);
+                return _context.SaveChangesAsync();
+            }
+        }
+
+
+        public async Task<Result> Delete(int roleId)
+        {
+            return await Get(roleId)
+                .Ensure(IsUnused, "This role is in use and connot be deleted")
```
Comment (mingazhev) [inline Api/AdministratorServices/AgentRolesService.cs:75]: Cannot

File: Api/AdministratorServices/AgentRolesService.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+            Task Add()
+            {
+                _context.AgentRoles.Add(roleInfo.ToAgentRole());
+                return _context.SaveChangesAsync();
+            }
+        }
+
+
+        public async Task<Result> Edit(int roleId, AgentRoleInfo roleInfo)
```
Comment (konstyefremov) [inline Api/AdministratorServices/AgentRolesService.cs:48]: Do we need to pass "int roleId" here if it's present inside "AgentRoleInfo roleInfo"?

File: Api/AdministratorServices/AgentRolesService.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+
+        public async Task<Result> Delete(int roleId)
+        {
+            return await Get(roleId)
+                .Ensure(IsUnused, "This role is in use and connot be deleted")
+                .Tap(Delete);
+
+
+            async Task<bool> IsUnused(AgentRole _)
```
Comment (konstyefremov) [inline Api/AdministratorServices/AgentRolesService.cs:79]: "AgentRole _" looks like can be safely removed.

File: Api/AdministratorServices/AgentRolesService.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                return _context.SaveChangesAsync();
+            }
+        }
+
+
+        private Result Validate(AgentRoleInfo roleInfo)
+            => GenericValidator<AgentRoleInfo>.Validate(v =>
+                {
+                    v.RuleFor(r => r.Name).NotEmpty();
+                    v.RuleFor(r => r.Permissions.ToList()).NotEmpty();
```
Comment (konstyefremov) [inline Api/AdministratorServices/AgentRolesService.cs:95]: Is it appropriate to use .ToList() here?

File: Api/Controllers/AdministratorControllers/AgentRolesController.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    public class AgentRolesController : BaseController
+    {
+        public AgentRolesController(IAgentRolesService agentRolesService)
+        {
+            _agentRolesService = agentRolesService;
+        }
+
+
+        /// <summary>
+        ///     Gets all pssible agent roles
```
Comment (konstyefremov) [inline Api/Controllers/AdministratorControllers/AgentRolesController.cs:25]: possible

File: Api/AdministratorServices/AgentRolesService.cs (hunk 3 above)
Comment (Raibo) [reply Api/AdministratorServices/AgentRolesService.cs:48]: Yes, we absolutely need, for two reasons:
1. API consistency, we gain access to an entity using it's Id in the endpoint path
2. Id in AgentRoleInfo instances are ignored. It is there only for responses, not requests.

File: Api/AdministratorServices/AgentRolesService.cs (hunk 1 above)
Comment (Raibo) [reply Api/AdministratorServices/AgentRolesService.cs]: done

File: Api/AdministratorServices/AgentRolesService.cs (hunk 4 above)
Comment (Raibo) [reply Api/AdministratorServices/AgentRolesService.cs]: This is due to FunctionalExtensions looks better when signature fits

File: Api/AdministratorServices/AgentRolesService.cs (hunk 5 above)
Comment (Raibo) [reply Api/AdministratorServices/AgentRolesService.cs]: fixed

**Comment by Raibo** [comment]
fixed and tested

**Review by mingazhev — APPROVED** [review]

**Pull request merged**